File: picture.cpp
/*
ID: wrwwctb1
TASK: picture
LANG: C++14
*/

#include <iostream>
#include <string>
#include <fstream>
#include <sstream>
#include <cassert>

#include <vector>
#include <deque>
#include <list>
#include <set> // has multiset
#include <map> // has multimap
#include <unordered_set> // has unordered_multiset
#include <unordered_map> // has unordered_multimap
#include <stack>
#include <queue> // has priority_queue

#include <algorithm>
#include <cmath>
#include <cfloat>
#include <ctime>
#include <climits>
#include <iomanip>
#include <cstring>

#define FOR( i, N ) for( int (i) = 0; (i) < (N); (i)++ )
//#define FOR_( i, start, endplusone ) for( int (i) = (start); (i) < (endplusone); (i)++ )
const double EPS = 1e-6;
const double PI = 3.141592653589793;
using namespace std;
// profile: 27196474








const string filename("picture");








int checkEvents(vector<vector<int>> & events){
    int tot = 0;
    vector<int> num(20000, 0);
    for (auto event: events){
        int x = event[0];
        int io = event[1];
        int ym = event[2];
        int yM = event[3];

        vector<int> prev(20000);
        copy(num.begin(), num.end(), prev.begin());

        if (io == 0){
            for (int y = ym+10000; y < yM+10000; y++){
                num[y] += 1;
            }
        }else{
            for (int y = ym+10000; y < yM+10000; y++){
                num[y] -= 1;
            }
        }
        FOR (y, 20000){
            if (!prev[y] && num[y]){
                tot += 1;
            }else if(prev[y] && !num[y]){
                tot += 1;
            }
        }
    }
    return tot;
}


int main(int argc, char **argv){

    string filenameo = filename + ".out";
    string filenamei = filename + ".in";
    ofstream fout(filenameo.c_str());
    ifstream fin(filenamei.c_str());






    int N;
    fin >> N;
    vector<vector<int>> rects(N, vector<int>(4));
    FOR (n, N){
        fin >> rects[n][0]
            >> rects[n][1]
            >> rects[n][2]
            >> rects[n][3];
    }

    vector<vector<int>> xevents(2*N, vector<int>(4));
    vector<vector<int>> yevents(2*N, vector<int>(4));
    FOR (i, N){
        int xm = rects[i][0];
        int ym = rects[i][1];
        int xM = rects[i][2];
        int yM = rects[i][3];
        xevents[2*i][0] = xm;
        xevents[2*i][1] = 0;
        xevents[2*i][2] = ym;
        xevents[2*i][3] = yM;
        xevents[2*i+1][0] = xM;
        xevents[2*i+1][1] = 1;
        xevents[2*i+1][2] = ym;
        xevents[2*i+1][3] = yM;
        yevents[2*i][0] = ym;
        yevents[2*i][1] = 0;
        yevents[2*i][2] = xm;
        yevents[2*i][3] = xM;
        yevents[2*i+1][0] = yM;
        yevents[2*i+1][1] = 1;
        yevents[2*i+1][2] = xm;
        yevents[2*i+1][3] = xM;
    }

    sort(xevents.begin(), xevents.end());
    sort(yevents.begin(), yevents.end());

    int totx = checkEvents(xevents);
    int toty = checkEvents(yevents);

    cout << " " << totx + toty << endl;
    fout << totx + toty << endl;





    //system("pause");
    return 0;
}

File: picture.py
"""
ID: wrwwctb1
LANG: PYTHON3
TASK: picture
5.5.1

key idea
x, y considered separately. for each dir, record exposed side wall height
"""

import os, sys, re, time
from collections import Counter, deque, defaultdict
from queue import Queue
from copy import copy, deepcopy
from itertools import combinations, permutations, accumulate, \
                      combinations_with_replacement
from functools import lru_cache, cmp_to_key, partial as functools_partial
from heapq import heappush, heappop, nlargest, nsmallest
from bisect import bisect_left, bisect_right
from math import ceil, floor, factorial, gcd, modf, log, log2, log10, sqrt, \
         pi, sin, cos, tan, asin, acos, atan, atan2, hypot, erf, erfc, inf, nan
# sys.setrecursionlimit(5782)
# python -m cProfile -s time ha.py

print = functools_partial(print, flush=True)
input = lambda: fin.readline().strip('\n')

def printwrite(string):
    print(string)
    fout.write(str(string) + '\n')

filename = 'picture'
fin = open(filename + '.in', 'r')
fout = open(filename + '.out', 'w')












def checkEvents(events):
    tot = 0
    num = [0] * (10000 + 1 + 9999) # num[0] = # of coverage at loc: -10000
    for x, io, ym, yM in events:
        prev = copy(num)
        if io == 0:
            for y in range(ym+10000, yM+10000):
                num[y] += 1
        else:
            for y in range(ym+10000, yM+10000):
                num[y] -= 1

        for y in range(20000):
            if not prev[y] and num[y]:
                tot += 1
            elif prev[y] and not num[y]:
                tot += 1

    return tot

N = int(input())
rects = []
for n in range(N):
    line = list(map(int, input().split()))
    rects.append(line)

# pre process

xevents = []
yevents = []
for xm, ym, xM, yM in rects:
    xevents.append([xm, 0, ym, yM]) # 0: into a rect
    xevents.append([xM, 1, ym, yM]) # 1: leaving a rect
    yevents.append([ym, 0, xm, xM])
    yevents.append([yM, 1, xm, xM])


xevents.sort()
yevents.sort()

totx = checkEvents(xevents)
toty = checkEvents(yevents)

printwrite(totx + toty)
















fin.close()
fout.close()
del print, input

